core/server/worker/src/main/java/alluxio/worker/netty/DataServerBlockReadHandler.java

```diff
@@ -83,62 +126,119 @@ public void close() throws IOException {
   protected void initializeRequest(Protocol.ReadRequest request) throws Exception {
     mRequest = new BlockReadRequestInternal(request);
   }
 
   @Override
-  protected DataBuffer getDataBuffer(Channel channel, long offset, int len) throws IOException {
-    LocalFileBlockReader blockReader =
-        (LocalFileBlockReader) ((BlockReadRequestInternal) mRequest).mBlockReader;
-    Preconditions.checkArgument(blockReader.getChannel() instanceof FileChannel,
-        "Only FileChannel is supported!");
-    switch (mTransferType) {
-      case MAPPED:
-        ByteBuf buf = channel.alloc().buffer(len, len);
+  protected DataBuffer getDataBuffer(Channel channel, long offset, int len) throws Exception {
+    openBlock(channel);
+    BlockReader blockReader = ((BlockReadRequestInternal) mRequest).mBlockReader;
+
+    if (mTransferType == TRANSFER && !(blockReader instanceof UnderFileSystemBlockReader)) {
+      return new DataFileChannel(new File(((LocalFileBlockReader) blockReader).getFilePath()),
+          offset, len);
+    } else {
+      ByteBuf buf = channel.alloc().buffer(len, len);
+      try {
+        while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
+        }
+        return new DataNettyBufferV2(buf);
+      } catch (Throwable e) {
+        buf.release();
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Opens the UFS block if it is not open.
+   *
+   * @param channel the netty channel
+   * @throws Exception if it fails to open the UFS block
+   */
+  private void openBlock(Channel channel) throws Exception {
+    BlockReadRequestInternal request = (BlockReadRequestInternal) mRequest;
+    if (request.mBlockReader != null) {
+      return;
+    }
+
+    int retryInterval = Constants.SECOND_MS;
+    RetryPolicy retryPolicy = new TimeoutRetry(UFS_BLOCK_OPEN_TIMEOUT_MS, retryInterval);
+
+    // TODO(calvin): Update the locking logic so this can be done better
+    if (request.mPromote) {
+      try {
+        mWorker.moveBlock(request.mSessionId, request.mId, mStorageTierAssoc.getAlias(0));
+      } catch (Exception e) {
+        LOG.warn("Failed to promote block {}: {}", request.mId, e.getMessage());
+      }
+    }
+
+    do {
+      long lockId;
+      if (request.isPersisted()) {
+        lockId = mWorker.lockBlockNoException(request.mSessionId, request.mId);
+      } else {
+        lockId = mWorker.lockBlock(request.mSessionId, request.mId);
+      }
+      if (lockId != BlockLockManager.INVALID_LOCK_ID) {
         try {
-          FileChannel fileChannel = (FileChannel) blockReader.getChannel();
-          Preconditions.checkState(fileChannel.position() == offset);
-          while (buf.writableBytes() > 0
-              && buf.writeBytes(fileChannel, buf.writableBytes()) != -1) {
-          }
-          return new DataNettyBufferV2(buf);
-        } catch (Throwable e) {
-          buf.release();
+          request.mBlockReader = mWorker.readBlockRemote(request.mSessionId, request.mId, lockId);
```

Comment:
this method `readBlockRemote` is meant to serve block read from remote client via netty. with domain socket, I guess the "remote" part is no more meaningful?